confidential-data-hub/hub/src/auth/kbs.rs

```diff
@@ -4,84 +4,93 @@
 use tokio::fs;
 
 use crate::{hub::Hub, Error, Result};
 
-impl Hub {
-    pub(crate) async fn init_kbs_resources() -> Result<()> {
-        let cmdline = fs::read_to_string("/proc/cmdline")
-            .await
-            .map_err(|e| Error::InitializationFailed(format!("read kernel cmdline failed: {e}")))?;
-        let kbs_resources = cmdline
-            .split_ascii_whitespace()
-            .find(|para| para.starts_with("cdh.kbs_resources="))
-            .unwrap_or("cdh.kbs_resources=")
-            .strip_prefix("cdh.kbs_resources=")
-            .expect("must have one")
-            .split(',')
-            .filter(|s| !s.is_empty())
-            .collect::<Vec<&str>>();
+/// This directory is used to store all the kbs resources get by CDH's init
+/// function, s.t. `[[Credential]]` sections in the config.toml file.
+pub const KBS_RESOURCE_STORAGE_DIR: &str = "/run/confidential-containers/cdh";
 
+impl Hub {
+    pub(crate) async fn init_kbs_resources(&self) -> Result<()> {
         let mut kbs_client = KbcClient::new()
             .await
             .map_err(|e| Error::InitializationFailed(format!("kbs client creation failed: {e}")))?;
 
-        // for each `kbs://...::/...` string
-        for resource_pair in kbs_resources {
-            let s = resource_pair.split("::").collect::<Vec<&str>>();
-            if s.len() != 2 {
-                return Err(Error::InitializationFailed(format!(
-                    "illegal `cdh.kbs_resources` item from kernel commandline: {resource_pair}"
-                )));
-            }
-
-            let mut target_path = PathBuf::new();
-            target_path.push(s[1]);
-            let parent_dir = target_path
-                .parent()
-                .ok_or(Error::InitializationFailed(format!(
-                    "illegal `cdh.kbs_resources` path item from kernel commandline: {}",
-                    s[1]
-                )))?;
-            fs::create_dir_all(parent_dir).await.map_err(|e| {
+        fs::create_dir_all(KBS_RESOURCE_STORAGE_DIR)
+            .await
+            .map_err(|e| {
                 Error::InitializationFailed(format!(
-                    "get kbs resource failed when creating dir: {e}"
+                    "Create {KBS_RESOURCE_STORAGE_DIR} dir failed {e:?}."
                 ))
             })?;
 
-            let contents = kbs_client
-                .get_secret(s[0], &Annotations::default())
+        for (k, v) in &self.credentials {
+            let content = kbs_client
+                .get_secret(v, &Annotations::default())
                 .await
                 .map_err(|e| {
                     Error::InitializationFailed(format!("kbs client get resource failed: {e}"))
                 })?;
 
-            fs::write(target_path, contents).await.map_err(|e| {
-                Error::InitializationFailed(format!(
-                    "kbs client get resource failed when writing to file: {e}"
-                ))
+            if !is_path_valid(k) {
```

Comment:
This check seems OK, but its placement seems a bit buried (in the middle of a key-value loop after a secret's already been fetched). Maybe `self.credentials` should be sanitized before this point.

Reply:
Sounds good. Let me fix this

Reply:
I move the whole check logic before creating KBS client.